core/pkg/artifacts/saver.go

```diff
@@ -295,10 +329,164 @@ func newUploadTask(fileInfo serverFileResponse, localPath string) *filetransfer.
+		return nil, nil
+	}
+	if fileSize > S3MaxMultiUploadSize {
+		return nil, fmt.Errorf("file size exceeds maximum S3 object size: %v", fileSize)
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	partsInfo := []gql.UploadPartsInput{}
+	partNumber := int64(1)
+	buffer := make([]byte, getChunkSize(fileSize))
+	for {
+		bytesRead, err := file.Read(buffer)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+		if bytesRead == 0 {
+			break
+		}
+		partsInfo = append(partsInfo, gql.UploadPartsInput{
+			PartNumber: partNumber,
+			HexMD5:     utils.ComputeHexMD5(buffer[:bytesRead]),
+		})
+		partNumber++
+	}
+	return partsInfo, nil
+}
+
+func (as *ArtifactSaver) uploadMultipart(
+	path string,
+	fileInfo serverFileResponse,
+	partData []gql.UploadPartsInput,
+	doneChan chan<- uploadResult,
+) {
+	statInfo, err := os.Stat(path)
+	if err != nil {
+		doneChan <- uploadResult{name: fileInfo.name, err: err}
+		return
+	}
+	chunkSize := getChunkSize(statInfo.Size())
+
+	type partResponse struct {
+		partNumber int64
+		task       *filetransfer.Task
+	}
+
+	wg := sync.WaitGroup{}
+	subChan := make(chan partResponse, len(partData))
+	// TODO: add mid-upload cancel.
+
+	var contentType string
+	for _, h := range fileInfo.uploadHeaders {
+		if strings.HasPrefix(h, "Content-Type:") {
+			contentType = strings.TrimPrefix(h, "Content-Type:")
+			break
+		}
+	}
+	if contentType == "" {
+		err := fmt.Errorf("content-type header is required for multipart uploads")
+		doneChan <- uploadResult{name: fileInfo.name, err: err}
+		return
+	}
+
+	partInfo := fileInfo.multipartUploadInfo
+	for i, part := range partInfo {
+		task := newUploadTask(fileInfo, path)
+		task.Url = part.UploadUrl
+		task.Offset = int64(i) * chunkSize
+		remainingSize := statInfo.Size() - task.Offset
+		task.Size = min(remainingSize, chunkSize)
+		b64md5, err := utils.HexToB64(partData[i].HexMD5)
+		if err != nil {
+			panic(err)
+		}
+		task.Headers = []string{
+			"Content-Md5:" + b64md5,
+			"Content-Length:" + strconv.FormatInt(task.Size, 10),
+			"Content-Type:" + contentType,
+		}
+		task.SetCompletionCallback(func(t *filetransfer.Task) {
+			subChan <- partResponse{partNumber: partData[i].PartNumber, task: t}
+			wg.Done()
+		})
+		wg.Add(1)
+		as.FileTransferManager.AddTask(task)
+	}
+
+	go func() {
+		wg.Wait()
+		close(subChan)
+	}()
+
+	partEtags := make([]gql.UploadPartsInput, len(partData))
+
+	for t := range subChan {
+		err := t.task.Err
+		if err == nil && t.task.Response == nil {
+			err = fmt.Errorf("no response in task %v", t.task.Name)
+		}
+		etag := ""
+		if err == nil && t.task.Response != nil {
+			etag = t.task.Response.Header.Get("ETag")
+			if etag == "" {
+				err = fmt.Errorf("no ETag in response %v", t.task.Response.Header)
+			}
+		}
+		if err != nil {
+			doneChan <- uploadResult{name: fileInfo.name, err: err}
+			return
+		}
+		partEtags[t.partNumber-1] = gql.UploadPartsInput{
```

Comment:
This is dangerous because you risk a panic based on the server response, so a bug in the server can break a user's run with no useful feedback to the user. It would be a better user experience if the artifacts failed to upload.

You should gather the inputs in a `map[int]gql.UploadPartsInput` and turn it into a slice later:

```
for i := 0; i < len(partNumberToInput); i++ {
    input, ok := partNumberToInput[i+1]
    if !ok {
        // fail the upload?
        ...
    }
    ...
}
```

Reply:
Thanks for noticing that! Rather than moving them from one data structure to another I'm just doing the part number checking at entry time before adding them to the slice.